### PR TITLE
Add App Engine system tests

### LINKED ISSUE #41: Create system tests for GAE credentials

### PR DESCRIPTION
**Pull request by jonparrott**
~~Has #58 as a diffbase.~~

Resolves #41, progress towards #55.

### CODE REVIEW

File: system_tests/app_engine/app/main.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+
+@contextlib.contextmanager
+def capture():
+    """Context manager that captures stderr and stdout."""
+    oldout, olderr = sys.stdout, sys.stderr
+    try:
+        out = StringIO()
+        sys.stdout, sys.stderr = out, out
+        yield out
```
Comment (dhermes) [inline system_tests/app_engine/app/main.py]: Why not keep them separated?

File: system_tests/app_engine/app/main.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+def test_default():
+    credentials, project_id = google.auth.default()
+
+    assert isinstance(credentials, app_engine.Credentials)
+    assert project_id == app_identity.get_application_id()
+
+
+@contextlib.contextmanager
+def capture():
+    """Context manager that captures stderr and stdout."""
```
Comment (dhermes) [inline system_tests/app_engine/app/main.py]: 1. You should add why you can't use `py.test` on GAE in this docstring
1. You copy-pasta this from `py.test` source?

File: system_tests/app_engine/app/main.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+class MainHandler(webapp2.RequestHandler):
+    def get(self):
+        self.response.headers['content-type'] = 'text/plain'
+
+        status, output = run_tests()
+
+        if not status:
+            self.response.status = 500
+
+        self.response.write(output)
```
Comment (dhermes) [inline system_tests/app_engine/app/main.py]: Oh I get it.

File: system_tests/app_engine/app/main.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+
+app = webapp2.WSGIApplication([
+    ('/', MainHandler)
```
Comment (dhermes) [inline system_tests/app_engine/app/main.py]: Cough-trailing-coughma

File: system_tests/app_engine/app/main.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    output = ''
+
+    with capture() as capsys:
+        try:
+            test_credentials()
+            test_default()
+            status = True
+        except Exception:
+            status = False
+            output = 'Stacktrace:\n{}\n'.format(traceback.format_exc())
```
Comment (dhermes) [inline system_tests/app_engine/app/main.py]: Don't you want to let them both fail and capture that? This way if `test_credentials()` fails `test_default()` won't run.

File: system_tests/app_engine/app/main.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    """
+    status = False
+    output = ''
+
+    with capture() as capsys:
+        try:
+            test_credentials()
+            test_default()
+            status = True
+        except Exception:
```
Comment (dhermes) [inline system_tests/app_engine/app/main.py]: Can't get a more specific exception? Isn't it just like an `AssertionError`? Or you want to keep your process from exploding?

File: system_tests/app_engine/app/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,3 @@
+urllib3
+# Relative path to google-auth-python's source.
+../../..
```
Comment (dhermes) [inline system_tests/app_engine/app/requirements.txt]: Ha. Fun.

File: system_tests/app_engine/test_app_engine.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+HERE = os.path.dirname(__file__)
+TEST_APP_DIR = os.path.join(HERE, 'app')
+TEST_APP_SERVICE = 'google-auth-system-tests'
+
+
+def vendor_app_dependencies():
+    """Vendors in the test application's third-party dependencies."""
+    subprocess.check_call(
+        ['pip', 'install', '-t', 'lib', '-r', 'requirements.txt'])
```
Comment (dhermes) [inline system_tests/app_engine/test_app_engine.py]: Holy moly this is happening.

Are there long forms of `-t` and `-r`?

File: system_tests/app_engine/test_app_engine.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+
+def deploy_app():
+    """Deploys the test application using gcloud."""
+    subprocess.check_call(
+        ['gcloud', 'app', 'deploy', '-q', 'app.yaml'])
```
Comment (dhermes) [inline system_tests/app_engine/test_app_engine.py]: Why `-q`? Why suppress this output but not the output of `pip`?

File: system_tests/app_engine/test_app_engine.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+    application_id = _cloud_sdk.get_project_id()
+    application_url = 'https://{}-dot-{}.appspot.com'.format(
+        TEST_APP_SERVICE, application_id)
+
+    yield application_url
+
+
+@pytest.mark.skipif(
+    'SKIP_APP_ENGINE_SYSTEM_TEST' in os.environ,
```
Comment (dhermes) [inline system_tests/app_engine/test_app_engine.py]: Make this env. var. a constant

File: system_tests/app_engine/test_app_engine.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    reason='Explicitly skipping App Engine system tests.')
+def test_live_application(app, http_request):
+    response = http_request(method='GET', url=app)
+    assert response.status == 200, response.data.decode('utf-8')
```
Comment (dhermes) [inline system_tests/app_engine/test_app_engine.py]: Why not two lines of asserts? What's the second one asserting? That the response isn't empty?

Also, if the response failed (`!= 200`) don't you want to print the response data?

File: system_tests/app_engine/app/appengine_config.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+# limitations under the License.
+
+from google.appengine.ext import vendor
+
+# Add any libraries installed in the "lib" folder.
+vendor.add('lib')
+
+
+# Patch os.path.expanduser. This should be fixed in GAE
+# versions released after Nov 2016.
```
Comment (jonparrott) [inline system_tests/app_engine/app/appengine_config.py]: I know. ☹️

File: system_tests/app_engine/app/main.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import contextlib
+import json
+import sys
+from StringIO import StringIO
```
Comment (jonparrott) [inline system_tests/app_engine/app/main.py]: It is not. Let's cross that bridge if it's ever built.

File: system_tests/app_engine/app/main.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    scoped_credentials = credentials.with_scopes([EMAIL_SCOPE])
+
+    scoped_credentials.refresh(None)
+
+    assert scoped_credentials.valid
+    assert scoped_credentials.token is not None
+
+    # Get token info and verify scope
+    url = _helpers.update_query(TOKEN_INFO_URL, {
+        'access_token': scoped_credentials.token
```
Comment (jonparrott) [inline system_tests/app_engine/app/main.py]: Done.

File: system_tests/app_engine/app/main.py [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    })
+    response = HTTP_REQUEST(url=url, method='GET')
+    token_info = json.loads(response.data.decode('utf-8'))
+
+    assert token_info['scope'] == EMAIL_SCOPE
```
Comment (jonparrott) [inline system_tests/app_engine/app/main.py]: There are other things you can verify but they aren't as deterministic as I'd like.

File: system_tests/app_engine/app/main.py (hunk 2 above)
Comment (jonparrott) [inline system_tests/app_engine/app/main.py]: 1) Added module-level docstring that describes this.

2) Nah, py.test's capsys is pretty complicated (it also supports capturing to a file descriptor).

File: system_tests/app_engine/app/main.py (hunk 1 above)
Comment (jonparrott) [inline system_tests/app_engine/app/main.py]: (You figured it out below, because they'll both to go the same spot and separating them breaks temporal cohesion)

File: system_tests/app_engine/app/main.py (hunk 6 above)
Comment (jonparrott) [inline system_tests/app_engine/app/main.py]: I would like to capture all errors and report them back to the test runner so that they are visible without needing to introspect the application's logs.

File: system_tests/app_engine/app/main.py (hunk 5 above)
Comment (jonparrott) [inline system_tests/app_engine/app/main.py]: I refactored a bit, let me know what you think.

File: system_tests/app_engine/app/main.py (hunk 4 above)
Comment (jonparrott) [inline system_tests/app_engine/app/main.py]: Done, but god help me if I ever add another handler here.

File: system_tests/app_engine/app/requirements.txt (hunk 7 above)
Comment (jonparrott) [inline system_tests/app_engine/app/requirements.txt]: The most™

File: system_tests/app_engine/test_app_engine.py (hunk 8 above)
Comment (jonparrott) [inline system_tests/app_engine/test_app_engine.py]: > Holy moly this is happening.

tumblr_inline_nzb6a3hkjg1rp5ay0_500

> Are there long forms of -t and -r

`-r` is well known, so I'll leave it as-is, I expanded `-t` to `--target`.

File: system_tests/app_engine/test_app_engine.py (hunk 9 above)
Comment (jonparrott) [inline system_tests/app_engine/test_app_engine.py]: `-q` doesnt' suppress output, it prevents it from asking to confirm. It's like `apt-get -y`.

File: system_tests/app_engine/test_app_engine.py (hunk 10 above)
Comment (jonparrott) [inline system_tests/app_engine/test_app_engine.py]: Done.

File: system_tests/app_engine/test_app_engine.py (hunk 11 above)
Comment (jonparrott) [inline system_tests/app_engine/test_app_engine.py]: `assert expr1, expr2` is [the same as](https://docs.python.org/2/reference/simple_stmts.html#the-assert-statement)

```python
if not expr1:
    raise AssertionError(expr2)
```

So what I'm doing here is asserting that status is 200, and if it's not, I'm raising an assertion error with the response data.

File: system_tests/app_engine/app/main.py [hunk 16]
```diff
@@ -70,26 +77,38 @@ def capture():
...
+def run_test_func(func):
+    with capture() as capsys:
+        try:
+            func()
+            return True, ''
+        except Exception as exc:
+            output = (
+                'Test {} failed: {}\n\n'
+                'Stacktrace:\n{}\n\n'
+                'Captured output:\n{}').format(
```
Comment (dhermes) [inline system_tests/app_engine/app/main.py]: Nittiest of nitpicks: Make this template string a global?

File: system_tests/app_engine/app/main.py [hunk 17]
```diff
@@ -70,26 +77,38 @@ def capture():
...
-            test_default()
-            status = True
-        except Exception:
-            status = False
-            output = 'Stacktrace:\n{}\n'.format(traceback.format_exc())
-
-    output += 'Captured output:\n{}'.format(capsys.getvalue())
+    tests = (test_credentials, test_default)
+
+    for test in tests:
```
Comment (dhermes) [inline system_tests/app_engine/app/main.py]: If you do decide to make another commit for my nitpick, while you're in there you could just say:

```python
for test in (test_credentials, test_default):
    ...
```

File: system_tests/app_engine/app/main.py (hunk 16 above)
Comment (jonparrott) [inline system_tests/app_engine/app/main.py]: Done.

File: system_tests/app_engine/app/main.py (hunk 17 above)
Comment (jonparrott) [inline system_tests/app_engine/app/main.py]: I prefer to keep it as a separate variable.
